pandas/core/indexes/base.py

```diff
@@ -687,7 +695,7 @@ def _cleanup(self):
     @cache_readonly
     def _engine(self):
         # property, for now, slow to look up
-        return self._engine_type(lambda: self._ndarray_values, len(self))
+        return self._engine_type(partial(_return_obj, self._ndarray_values), len(self))
```

Comment:
would this work if return_obj were defined within `_engine`?

Is return_obj going to be useful in other places too?

Reply:
You're right, the nested function / lambda is NOT the issue, but the problem is that `self` is bound to the `lambda`. I've added a more elegant solution.